src/qibo/transpiler/optimizer.py

```diff
@@ -66,3 +67,261 @@ def __call__(self, circuit: Circuit):
             else:
                 new.add(fgate)
         return new
+
+
+class InverseCancellation(Optimizer):
+    """
+    Identifies and removes pairs of adjacent gates from a quantum circuit.
+    """
+
+    def __init__(self):
+        self.inverse_gates = (
+            gates.H | gates.Y | gates.Z | gates.X | gates.CNOT | gates.CZ | gates.SWAP
+        )
+        self.pos_rm_inv_gate = []
+
+    def __call__(self, circuit: Circuit) -> Circuit:
+        self.__find_pos_rm(circuit)
+
+        if 0 == len(self.pos_rm_inv_gate) or circuit.gates_of_type(gates.FusedGate):
+            return circuit
+
+        tmp_new_circuit = circuit.copy(True)
+
+        new_circuit = circuit.__class__(**tmp_new_circuit.init_kwargs)
+        for i, gate in enumerate(tmp_new_circuit.queue):
+            if not (i in self.pos_rm_inv_gate):
+                new_circuit.add(gate)
+
+        return new_circuit
+
+    def __find_pos_rm(self, newCircuit: Circuit):
+        """
+        Identifies and marks pairs of inverse gates that can be removed from the circuit.
+
+        Args:
+             newCircuit: The Circuit.
+
+        """
+
+        previous_gates = [None] * newCircuit.nqubits
+
+        def update_previous_gates(qubits, idx, gate):
+            """Helper function to update previous gate tracking."""
+            for q in qubits:
+                previous_gates[q] = (idx, gate)
+
+        def clear_previous_gates(qubits):
+            """Helper to clear tracking for given qubits."""
+            for q in qubits:
+                previous_gates[q] = None
```

Comment:
I advise to now define functions inside other functions unless it is strictly necessary